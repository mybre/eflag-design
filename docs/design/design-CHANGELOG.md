---
title: 更新日志
order: 6
group: 基础组件
---

`@oceanbase/design` 严格遵循 [Semantic Versioning 2.0.0](http://semver.org/lang/zh-CN/) 语义化版本规范。

---

## 0.2.32

`2023-12-14`

- 🔥 Space 和 Grid 组件的间距样式兼容 Chrome 84 以下的浏览器。[#344](https://github.com/oceanbase/oceanbase-design/pull/344) [@wdyea-ya](https://github.com/wdyea-ya)
- 🌈 更新默认主题的功能色板，包括 Design Token 和 less 变量。[#354](https://github.com/oceanbase/oceanbase-design/pull/354)
- Tag
  - 🆕 Tag 新增 `ellipsis` 属性，用于配置内容溢出时的省略和 Tooltip 提示。[#361](https://github.com/oceanbase/oceanbase-design/pull/361)
  - 🐞 修复 Tag 字体大小不正确的问题。[#360](https://github.com/oceanbase/oceanbase-design/pull/360)
- 🐞 修复 Select 在多选模式和 `large` & `small` 尺寸下，选中项缺少边框样式的问题。[#332](https://github.com/oceanbase/oceanbase-design/pull/332) [@wdyea-ya](https://github.com/wdyea-ya)
- 💄 优化 Button 样式，包括更新主按钮的渐变色，并去除 `box-shadow` 阴影。[#352](https://github.com/oceanbase/oceanbase-design/pull/352)
- TypeScript
  - 🤖 导出 SpaceProps、SpaceSize、ColProps、ColSize 和 RowProps 的类型定义。[#344](https://github.com/oceanbase/oceanbase-design/pull/344)

## 0.2.31

`2023-12-08`

- 🆕 ConfigProvider 新增 `styleProviderProps` 属性，一般用于配置 StyleProvider 的 `hashPriority` 和 `transformers` 属性实现样式降级，来兼容 Chrome 88 以下的低版本浏览器。[#343](https://github.com/oceanbase/oceanbase-design/pull/343)
- 🐞 修复 Drawer 的分隔阴影没有跟随内容和窗口尺寸变化实时更新的问题。[#337](https://github.com/oceanbase/oceanbase-design/pull/337)
- 🐞 修复 Form `hideRequiredMark` 优先级低于 ConfigProvider `form.requiredMark` 的问题。[#349](https://github.com/oceanbase/oceanbase-design/pull/349)

## 0.2.30

`2023-11-30`

- 🌈 更新功能色板，包括 Design Token 和 less 变量。
- 🔥 实现新版 Drawer 的样式和交互。[#319](https://github.com/oceanbase/oceanbase-design/pull/319)
  - 优化页头、内容区和页脚的间距。
  - 当内容区高度大于抽屉高度时，页脚置底展示；当内容区高度小于抽屉高度时，页脚跟随内容展示。
  - 内容区滚动时，动态设置页头和页脚的阴影，实现和内容区的分隔效果。
  - 页脚操作区的主按钮位置居左。
- 📢 Form 默认开启 `requiredMark: optional` 可选样式。[#312](https://github.com/oceanbase/oceanbase-design/pull/312)
- 📢 Table 和 List 默认开启 `pagination.hideOnSinglePage`，即只有一页数据时会隐藏分页器。[#330](https://github.com/oceanbase/oceanbase-design/pull/330)
- 🆕 ConfigProvider 支持全局配置 `hideOnSinglePage`。[#330](https://github.com/oceanbase/oceanbase-design/pull/330)
- 🆕 Segmented `options` 选项新增 `ellipsis` 属性，用于配置内容溢出时的省略和 Tooltip 提示。[#227](https://github.com/oceanbase/oceanbase-design/pull/227) [@TianWuwt](https://github.com/TianWuwt)
- Descriptions
  - 🆕 Descriptions `items` 新增 `contentProps` 属性，用于设置省略、编辑、复制等内容属性，仅无边框模式下生效。[#329](https://github.com/oceanbase/oceanbase-design/pull/329)
  - 💄 Descriptions 垂直布局下默认去掉 `colon` 冒号。[#328](https://github.com/oceanbase/oceanbase-design/pull/328)
- Badge
  - 🐞 修复开启状态图标时间距样式不生效的问题。[#300](https://github.com/oceanbase/oceanbase-design/pull/300)
  - 💄 状态文本的默认字体颜色，会继承父元素的设置，而不总是 `token.colorText`，便于和其他组件组合使用。[#322](https://github.com/oceanbase/oceanbase-design/pull/322)
- Card
  - 🐞 修复 Card 分隔线可能会被内容区遮挡的问题。[#326](https://github.com/oceanbase/oceanbase-design/pull/326)
  - 💄 去掉无边框内部卡片的阴影，优化嵌套卡片的样式效果。[#325](https://github.com/oceanbase/oceanbase-design/pull/325)
- 💄 Typography.Text 和 Typography.Paragraph 的默认字体颜色和行高，会继承父元素的设置，而不总是 `token.colorText` 和 `token.lineHeight`，便于和其他组件组合使用。[#321](https://github.com/oceanbase/oceanbase-design/pull/321)
- TypeScript
  - 🤖 导出 AlertProps、CardProps、ConfigProviderProps、DescriptionsItemProps、FormItemProps、ModalProps、ModalProgressProps、DrawerProps、TableProps、TabsProps、TagProps、TooltipProps、SpinProps 和 BadgeProps 等扩展组件的 TS 类型。[#311](https://github.com/oceanbase/oceanbase-design/pull/311)

## 0.2.29

`2023-11-17`

- 🐞 修复 Drawer 部分样式不生效的问题。[#298](https://github.com/oceanbase/oceanbase-design/pull/298)

## 0.2.28

`2023-11-14`

- 🌈 新增主题预览和编辑器，便于主题调试和预览。[#287](https://github.com/oceanbase/oceanbase-design/pull/287)
- 💄 优化 Drawer 的标题、内容区和页脚样式，以符合设计规范。[#283](https://github.com/oceanbase/oceanbase-design/pull/283) [@Vanleehao](https://github.com/Vanleehao)

## 0.2.27

`2023-11-09`

- 🌈 更新中性色板，包括 Design Token 以及 less 主题变量。[#272](https://github.com/oceanbase/oceanbase-design/pull/272)
- 🔥 Drawer [#228](https://github.com/oceanbase/oceanbase-design/pull/228) [@Vanleehao](https://github.com/Vanleehao)
  - 🌈 定制主题和样式，符合 OceanBase Design 设计规范。
  - 🆕 新增 `footer` 属性，用于设置抽屉的底部内容，默认为 `取消` 和 `确定` 按钮。
  - 🆕 新增 `onOk` 和 `onCancel` 属性，用于设置 `取消` 和 `确定` 按钮的回调。
  - 🆕 新增 `cancelText` 和 `okText` 属性，用于设置 `取消` 和 `确定` 按钮的文字。
  - 🆕 新增 `okButtonProps` 属性，用于设置 `确定` 按钮的属性。
  - 🆕 新增 `confirmLoading` 属性，用于设置 `确定` 按钮的加载态。
- 🆕 全局设置 Spin 的加载指示符为 OceanBase 加载动画。[#273](https://github.com/oceanbase/oceanbase-design/pull/273)
- 🆕 Badge、Button、Card、Select、Table、Tag 和 Tooltip 支持转发 `ref`。[#259](https://github.com/oceanbase/oceanbase-design/pull/259) [@linhf123](https://github.com/linhf123)
- 💄 优化 Modal 确认对话框的内容区样式。[#275](https://github.com/oceanbase/oceanbase-design/pull/275)

## 0.2.26

`2023-11-03`

## 0.2.25

`2023-10-31`

- 🔥 新增 Tag 组件，优化 border 样式以弱化边框，内容超长支持自动省略，可通过 `ellipsis` 属性进行控制。[#113](https://github.com/oceanbase/oceanbase-design/pull/113) [@wdyea-ya](https://github.com/wdyea-ya)
- 🐞 修复 Tooltip 未兼容 `visible` 属性导致显示和隐藏无法受控的问题。[#231](https://github.com/oceanbase/oceanbase-design/pull/231)

## 0.2.24

`2023-10-26`

- Table
  - 🐞 修复 `columns` 为空时，Table 执行报错、导致页面白屏的问题。[#206](https://github.com/oceanbase/oceanbase-design/pull/206)
  - 💄 优化带边框的 Table 样式，去掉其底部多余的 border。[#207](https://github.com/oceanbase/oceanbase-design/pull/207)
