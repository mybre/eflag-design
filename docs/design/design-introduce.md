---
title: Eflag Design 介绍
order: 1
group:
  title: 基础组件
  order: 1
---

Eflag Design 是 Eflag 前端和设计团队出品的 UI 资产库，与 Ant Design 设计体系一脉相承。`@eflag/design` 是 `Eflag Design` 的 React 实现，基于 [antd](https://ant.design/) 做了大量样式定制、能力扩展和封装。

## ✨ 特性

- 🌈 从业务中来，提炼自 Eflag 各个产品的视觉和交互风格。
- 🎨 与 Ant Design 设计体系一脉相承，可从 antd 平滑迁移。
- ⚙️ 全链路开发和设计工具体系，与 Umi 框架无缝对接。
- 📦 开箱即用的高质量 React 组件。
- 🛡 使用 TypeScript 开发，提供完整的类型定义文件。
- 🌍 国际化语言支持。

## 支持环境

- 现代浏览器和 IE11 及以上。
- Electron

<embed src="./design-quickstart.md#L8-L100000"></embed>
