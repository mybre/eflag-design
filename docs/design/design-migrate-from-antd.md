---
title: 从 antd 迁移
order: 5
group: 基础组件
---

OceanBase Design 与 Ant Design 设计体系一脉相承，完全继承 [Ant Design](https://ant.design) 的能力和 API。两者仅在组件主题、样式和扩展能力上存在差异，因此只需简单替换模块入口，即可从 Ant Design 平滑迁移至 OceanBase Design。

## 自动化迁移

- 使用 [@oceanbase/codemod](https://github.com/oceanbase/oceanbase-design/tree/master/packages/codemod) 可以从 `antd` 一键迁移到 `@oceanbase/design`:

```shell
# Run directly through npx
# `src` is the target directory or file that you want to transform.
npx -p @oceanbase/codemod codemod src
# options
# --disablePrettier   // disable prettier
```

## 手动迁移

```diff
- import { xxx } from 'antd';
- import { xxx } from 'antd/es/button';
+ import { xxx } from '@oceanbase/design';
+ import { xxx } from '@oceanbase/design/es/button';
```
