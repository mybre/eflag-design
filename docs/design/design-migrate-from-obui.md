---
title: 从 @alipay/ob-ui 迁移
order: 6
group: 基础组件
---

内部包 `@alipay/ob-ui` 由于包含多种类型的组件，不便于管理和使用。因此在对外开源时，将其拆分成了 `@oceanbase/design` 基础组件库和 `@oceanbase/ui` 业务组件库，且有一定的不兼容变更，建议使用我们提供的自动化工具进行迁移。

## 自动化迁移

- 使用 [@oceanbase/codemod](https://github.com/oceanbase/oceanbase-design/tree/master/packages/codemod) 可以从 `@alipay/ob-ui` 一键迁移到 `@oceanbase/design`:

```shell
# Run directly through npx
# `src` is the target directory or file that you want to transform.
npx -p @oceanbase/codemod codemod src
# options
# --disablePrettier   // disable prettier
```
