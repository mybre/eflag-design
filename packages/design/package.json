{
  "name": "@eflag/design",
  "version": "0.2.22",
  "description": "The Design System of eflag",
  "keywords": [
    "eflag",
    "design",
    "ui",
    "react",
    "component"
  ],
  "homepage": "https://github.com/oceanbase/oceanbase-design/packages/design",
  "repository": {
    "type": "git",
    "url": "git@github.com:oceanbase/design.git"
  },
  "publishConfig": {
    "registry": "https://registry.npmjs.org",
    "access": "public"
  },
  "main": "lib/index.js",
  "unpkg": "dist/design.min.js",
  "module": "es/index.js",
  "typings": "es/index.d.ts",
  "files": [
    "es",
    "lib",
    "dist",
    "package.json",
    "README.md",
    "LICENSE"
  ],
  "scripts": {
    "build": "father build",
    "postbuild": "cp src/style/reset.css dist/"
  },
  "dependencies": {
    "@ant-design/cssinjs": "^1.17.2",
    "@eflag/icons": "workspace:^",
    "@eflag/util": "workspace:^",
    "ahooks": "^2.10.14",
    "antd": "^5.10.1",
    "classnames": "^2.3.2",
    "lodash": "^4.17.21",
    "lottie-web": "^5.12.2",
    "prop-types": "^15.8.1",
    "rc-util": "^5.38.0",
    "react-sticky-mouse-tooltip": "^0.0.1"
  },
  "devDependencies": {
    "@babel/cli": "^7.23.0",
    "@babel/preset-env": "^7.23.2",
    "antd-token-previewer": "^1.1.0"
  },
  "peerDependencies": {
    "react": ">=16.9.0",
    "react-dom": ">=16.9.0"
  }
}