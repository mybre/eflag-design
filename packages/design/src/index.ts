// 引入 antd/dist/reset.css，以重置基本样式，保证原生元素遵从 antd 规范样式
// ref: https://ant.design/docs/react/migration-v5-cn#%E6%8A%80%E6%9C%AF%E8%B0%83%E6%95%B4
import 'antd/dist/reset.css';

export * from 'antd';
export { version } from '../package.json';

export { default as Alert } from './alert';
export type { AlertProps } from './alert';

export { default as Button } from './button';
export type { ButtonProps } from './button';

export { default as Space } from './space';
export type { SpaceProps, SpaceSize } from './space';

export { Col, Row } from './grid';
export type { ColProps, RowProps, ColSize } from './grid';

export { default as Typography } from './typography';
export type { TypographyProps } from './typography';

export { default as Card } from './card';
export type { CardProps } from './card';

export { default as ConfigProvider } from './config-provider';
export type { ConfigProviderProps } from './config-provider';

export { default as Descriptions } from './descriptions';
export type {
  DescriptionsProps,
  DescriptionsItemType,
  DescriptionsItemProps,
} from './descriptions';

export { default as Form } from './form';
export type { FormProps, FormItemProps } from './form';

export { default as Modal } from './modal';
export type { ModalProps, ModalFuncProps, ModalProgressProps } from './modal';

export { default as Drawer } from './drawer';
export type { DrawerProps } from './drawer';

export { message, notification, token } from './static-function';

export { default as List } from './list';
export type { ListProps } from './list';

export { default as Table } from './table';
export type { TableProps } from './table';

export { default as Tabs } from './tabs';
export type { TabsProps } from './tabs';

export { default as Tag } from './tag';
export type { TagProps } from './tag';

export { default as Select } from './select';
export type { SelectProps } from './select';

export { default as Tooltip } from './tooltip';
export type { TooltipProps } from './tooltip';

export { default as Segmented } from './segmented';
export type { SegmentedProps } from './segmented';

export { default as Breadcrumb } from './breadcrumb';
export type { BreadcrumbProps, BreadcrumbItemProps } from './breadcrumb';

export { default as Spin } from './spin';
export type { SpinProps } from './spin';

export { default as Badge } from './badge';
export type { BadgeProps } from './badge';

export { default as Lottie } from './lottie';
export type { LottieProps, LottieRef } from './lottie';

export { default as theme } from './theme';

import theme from './theme';

const { useToken } = theme;
export { useToken };

export type { PresetStatusColorType } from 'antd/es/_util/colors';
