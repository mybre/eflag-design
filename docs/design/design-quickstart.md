---
title: 快速上手
order: 2
group: 基础组件
---

## 使用 CodeSandbox 快速创建

- 可以使用我们提供的 [CodeSandbox 模板](https://codesandbox.io/s/oceanbase-design-reproduction-template-k26fm5) 快速创建和预览。

<iframe src="https://codesandbox.io/embed/oceanbase-design-reproduction-template-k26fm5?fontsize=14&hidenavigation=1&theme=dark"
     style="width:100%; height:500px; border:0; border-radius: 4px; overflow:hidden;"
     title="@eflag/design reproduction template"
     allow="accelerometer; ambient-light-sensor; camera; encrypted-media; geolocation; gyroscope; hid; microphone; midi; payment; usb; vr; xr-spatial-tracking"
     sandbox="allow-forms allow-modals allow-popups allow-presentation allow-same-origin allow-scripts"
   ></iframe>

## 在项目中使用

### 安装

```bash
$ npm i @eflag/design --save
```

### 示例

```jsx | pure
import { Table } from '@eflag/design';

const App = () => {
  const dataSource = [
    {
      key: '1',
      name: '胡彦斌',
      age: 32,
      address: '西湖区湖底公园1号',
    },
    {
      key: '2',
      name: '胡彦祖',
      age: 42,
      address: '西湖区湖底公园1号',
    },
    {
      key: '3',
      name: '胡彦斌',
      age: 32,
      address: '西湖区湖底公园1号',
    },
    {
      key: '4',
      name: '胡彦祖',
      age: 42,
      address: '西湖区湖底公园1号',
    },
    {
      key: '5',
      name: '胡彦斌',
      age: 32,
      address: '西湖区湖底公园1号',
    },
    {
      key: '6',
      name: '胡彦祖',
      age: 42,
      address: '西湖区湖底公园1号',
    },
    {
      key: '7',
      name: '胡彦斌',
      age: 32,
      address: '西湖区湖底公园1号',
    },
    {
      key: '8',
      name: '胡彦祖',
      age: 42,
      address: '西湖区湖底公园1号',
    },
  ];

  const columns = [
    {
      title: '姓名',
      dataIndex: 'name',
      key: 'name',
    },
    {
      title: '年龄',
      dataIndex: 'age',
      key: 'age',
    },
    {
      title: '住址',
      dataIndex: 'address',
      key: 'address',
    },
  ];

  return <Table columns={columns} dataSource={dataSource} />;
};
```
