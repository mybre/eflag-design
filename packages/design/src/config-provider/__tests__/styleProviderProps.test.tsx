import React, { useContext } from 'react';
import { render } from '@testing-library/react';
import { ConfigProvider } from '@oceanbase/design';
import { legacyLogicalPropertiesTransformer } from '@ant-design/cssinjs';
import StyleContext from '@ant-design/cssinjs/es/StyleContext';

describe('ConfigProvider styleProviderProps', () => {
  it('ConfigProvider styleProviderProps.hashPriority', () => {
    const Child1 = () => {
      const { hashPriority, transformers } = useContext(StyleContext);
      expect(hashPriority).toBe('low');
      expect(transformers).toBeUndefined();
      return <div />;
    };
    const Child2 = () => {
      const { hashPriority, transformers } = useContext(StyleContext);
      expect(hashPriority).toBe('low');
      expect(transformers).toBeUndefined();
      return <div />;
    };
    const Child3 = () => {
      const { hashPriority, transformers } = useContext(StyleContext);
      expect(hashPriority).toBe('high');
      expect(transformers.length).toBe(1);
      return <div />;
    };
    const { asFragment } = render(
      <ConfigProvider>
        <Child1 />
        <ConfigProvider>
          <Child2 />
        </ConfigProvider>
        <ConfigProvider
          styleProviderProps={{
            hashPriority: 'high',
            transformers: [legacyLogicalPropertiesTransformer],
          }}
        >
          <Child3 />
        </ConfigProvider>
      </ConfigProvider>
    );
    expect(asFragment().firstChild).toMatchSnapshot();
  });
});
