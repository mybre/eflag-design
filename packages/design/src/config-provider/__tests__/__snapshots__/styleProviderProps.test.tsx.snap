// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigProvider styleProviderProps ConfigProvider styleProviderProps.hashPriority 1`] = `
<div
  class="ant-app"
>
  <div />
  <div
    class="ant-app"
  >
    <div />
  </div>
  <div
    class="ant-app"
  >
    <div />
  </div>
</div>
`;
